File: Java/soln-backtracking-problems/subsets-ii.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/subsets-ii/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {    
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> subsets=new ArrayList<>();
        generateSubsets(0,nums, new ArrayList<Integer>(),subsets);
        return subsets;
        
    }
    
    public void generateSubsets(int index, int[] nums, List<Integer> cur, List<List<Integer>> subsets){
        subsets.add(new ArrayList<>(cur));
        for(int i=index;i<nums.length;i++){
            if(i>index && nums[i]==nums[i-1])
                continue;
            cur.add(nums[i]);
            generateSubsets(i+1,nums,cur,subsets);
            cur.remove(cur.size()-1);
        }
    }
}

File: C++/soln-backtracking-problems/subsets-ii.cpp
/**
 @author Farheen Bano

 Date 13-08-2021
  
 Reference-
 https://leetcode.com/problems/subsets-ii/
*/

class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> result;
        vector<int> cur;
        generateSubsets(0,nums,cur,result);
        return result;
    }
    
     void generateSubsets(int index, vector<int>& nums, vector<int>& cur, vector<vector<int>>& result){        
        result.push_back(cur);

        for(int i=index;i<nums.size();i++){
            if(i>index && nums[i]==nums[i-1])
                continue;
            cur.push_back(nums[i]);
            generateSubsets(i+1,nums,cur,result);
            cur.pop_back();
        }
    }
};